### PR TITLE
Deallocate memory and enable ASAN for examples

### PR DESCRIPTION
**Pull request by nurmukhametov**
This change let us enable ASAN with leak sanitizer on examples. There are still a few deallocations detected by LSAN On LIT tests and runtime tests.

### CODE REVIEW

File: src/ctx.cpp [hunk 1]
```diff
@@ -506,8 +506,10 @@ void FunctionEmitContext::EndIf() {
 
     // 'uniform' ifs don't change the mask so we only need to restore the
     // mask going into the if for 'varying' if statements
-    if (ci->IsUniform() || bblock == nullptr)
```
Comment (dbabokin) [inline src/ctx.cpp:509]: Would making `ci` a `unique_ptr` be a better / more elegant solution? I.e. declare it as `std::unique_ptr<CFInfo> ci(popCFState());`.

What's your opinion?

File: src/module.cpp [hunk 2]
```diff
@@ -3186,6 +3187,8 @@ int Module::CompileAndOutput(const char *srcFile, Arch arch, const char *cpu, st
         delete g->target;
         g->target = nullptr;
 
+        delete targetMachine;
```
Comment (dbabokin) [inline src/module.cpp:3199]: A pointer to `llvm::TargetMachine` is owned by `Target`, why not deleting it in its destructor?

File: src/ctx.cpp (hunk 1 above)
Comment (e3m3) [reply src/ctx.cpp:509]: I think the main advantage would be in the case the assertion is tripped, the destructor will still be called.

Comment (e3m3) [reply src/ctx.cpp:509]: Actually, I think this may be an issue for the path where the predicate is false, since the pointer should remain undeleted according to the current implementation. With unique_ptr, it will go out of scope later and be destructed. Unique_ptr assumes ownership.

File: src/main.cpp [hunk 3]
```diff
@@ -244,7 +244,9 @@ class ArgFactory {
   private:
     char *AllocateString(std::string string) {
         int len = string.length();
-        char *ptr = new char[len + 1];
+        // We use malloc here because of strdup in lAddSingleArg
```
Comment (dbabokin) [inline src/main.cpp:247]: I'm not sure how is this related. Could you elaborate?

Comment (dbabokin) [reply src/main.cpp:247]: Ah, I see. You want to deallocate both of them in the same manner and it ha to be `free()`.

I think a better solution would be to use only C++ primitives to work with strings, but this that would require more changes. So this is ok.

File: src/ctx.cpp (hunk 1 above)
Comment (dbabokin) [reply src/ctx.cpp:509]: I don't understand the last comment. Both explicit delete and `uniquie_ptr` will not change the current logic in this function.

Comment (e3m3) [reply src/ctx.cpp:509]: I mean that as it is now, if `! (ci->IsUniform() || bblock == nullptr)`, there's no call to delete.

Comment (e3m3) [reply src/ctx.cpp:509]: Ah never mind, I was looking at the left diff.

Comment (nurmukhametov) [reply src/ctx.cpp:509]: I am not sure about `unique_ptr`.
The problem here is that `popCFState` doesn't really create or allocate `CFInfo` object. The lifetime of `CFInfo` objects is the following (let's consider `if` only):
- Created via allocation on heap in `CFInfo::GetIf`: https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/ctx.cpp#L145
- `CFInfo::GetIf` returns a pointer to a newly allocated object that stored in `controlFlowInfo` vector in two places:
https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/ctx.cpp#L495
and
https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/ctx.cpp#L499
- The pointer is stored there until being poped in `popCFState` https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/ctx.cpp#L3771
- Then this pointer is returned by `popCFState` and accessed, e.g., in `EndIf` here.
So, we may bind lifetime of `CFInfo` object inside `CFInfo::GetIf` (and similar functions) to stack allocated `unique_ptr` but not the whole.

File: src/module.cpp (hunk 2 above)
Comment (nurmukhametov) [reply src/module.cpp:3199]: Unfortunately, `Target` doesn't really own `llvm::TargetMachine *m_targetMachine`, see comment here: https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/ispc.h#L354

and example of usage after `Target` object deconstruction:
- `Target` created https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/module.cpp#L3247
- Ownership of `m_targetMachine` is passed to the `targetMachines` https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/module.cpp#L3260
- `Target` destroyed https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/module.cpp#L3321
- `targetMachine` is used after that here https://github.com/ispc/ispc/blob/f080d39c99decb9947e7fa5688a56773f841fe11/src/module.cpp#L3340

File: src/main.cpp (hunk 3 above)
Comment (nurmukhametov) [reply src/main.cpp:247]: Yes, exactly.
`argv` vector consisted of pointers to C string of two origins:
- real arguments of program from `Argv` of `main`
- heap allocated strings.
They can be mixed in order inside `argv` so I didn't find a easy way to distinguish them in order to deallocate the last ones.

I have ended up with the current fix that `strdup` the first ones and allocate the last ones with `malloc` to deallocate all of them with `free`.

File: src/ctx.cpp (hunk 1 above)
Comment (nurmukhametov) [reply src/ctx.cpp:509]: Bearing that in mind, I have fixed that to use `unique_ptr` instead explicit `delete ci` that is better especially when multiple exit points exist.

Comment (dbabokin) [reply src/ctx.cpp]: `unique_ptr` doesn't need to manage the whole lifetime from creation to destruction, managing it from popping to destruction is ok from philosophical point of view.

File: src/module.cpp (hunk 2 above)
Comment (dbabokin) [reply src/module.cpp:3199]: Ok, that's caused by messy way of doing multi-target compilation.

**Review by aneshlya — APPROVED** [review]

File: src/module.cpp (hunk 2 above)
Comment (dbabokin) [reply src/module.cpp:3199]: I've had a deeper look and I think this needs to be cleaned up. Specifically:
- `TargetMachine` really belongs to the `Target` and needs to be destroyed in its destructor.
- The cases when it escapes `Target` lifetime are not legitimate:
  - `targetMachines[]` array should be a bool array with a name "CompiledTargets" or something like that.
  - the use of `firstTargetMachine` should be turned into `g->target->GetTargetMachine()`, as for the dispatch module we create a new `Target` anyway and it creates a `TargetMachine` as well - and this one needs to be used.

I'll proceed with merging this PR, and will create an issue to follow up with the clean up explained here.

Comment (dbabokin) [reply src/module.cpp:3199]: I've created https://github.com/ispc/ispc/issues/2578 to follow up.

**Review by dbabokin — APPROVED** [review]
LGTM
